synapse/rest/admin/users.py

```diff
@@ -229,37 +229,50 @@ async def on_PUT(
         if not isinstance(deactivate, bool):
             raise SynapseError(400, "'deactivated' parameter is not of type boolean")
 
-        # convert into List[Tuple[str, str]]
+        # convert List[Dict[str, str]] into List[Tuple[str, str]]
         if external_ids is not None:
             new_external_ids = []
             for external_id in external_ids:
                 new_external_ids.append(
                     (external_id["auth_provider"], external_id["external_id"])
                 )
 
+        # convert List[Dict[str, str]] into List[Tuple[str, str]]
+        if threepids is not None:
+            new_threepids = []
+            for threepid in threepids:
+                new_threepids.append((threepid["medium"], threepid["address"]))
```

Comment:
why not construct these as sets?

```suggestion
        # convert List[Dict[str, str]] into Set[Tuple[str, str]]
        if threepids is not None:
            new_threepids = {
                (threepid["medium"], threepid["address"])
                for threepid in threepids
            }
```

Likewise `cur_threepids` below - then you can remove the `set()` operations.

Reply:
I just didn't think far enough. :(